### PR TITLE
[codex] Type compaction task stages

### PR DESCRIPTION
**Pull request by henrypark133**
## Summary

- Refactors Reborn compaction task internals into private typed stages: validated range, compaction input, and sanitized summary.
- Keeps `LoopCompactionPort`, `SystemInferencePort`, and `SessionThreadService` public boundaries unchanged.
- Moves input byte-cap accounting into the escaped append path while preserving raw-body security scanning before serialization.

## Validation

- `cargo fmt --check`
- `git diff --check`
- `cargo test -p ironclaw_loop_support`
- `cargo test -p ironclaw_loop_support --test compaction_task_contract`
- `cargo test -p ironclaw_loop_support --test thread_loop_support_contract`
- `cargo test -p ironclaw_agent_loop compaction`
- `cargo clippy -p ironclaw_loop_support --all-targets -- -D warnings`

## Review

- Boundary ownership subagent: clean.
- Code review subagent: no findings.

### CODE REVIEW

File: crates/ironclaw_loop_support/src/compaction_task.rs [hunk 1]
```diff
@@ -199,70 +225,94 @@ where
...
+        Ok(ValidatedCompactionRange {
+            thread_id: request.thread_id.clone(),
+            thread_scope,
+            start_sequence: start_exclusive.saturating_add(1),
+            end_sequence: request.drop_through_seq,
+            messages: validated_messages,
+        })
+    }
+
+    fn build_input(
```
Comment (henrypark133) [inline crates/ironclaw_loop_support/src/compaction_task.rs:275]: **Medium** — Double injection/leak scan: per-body scan then full-concatenated scan.

In build_input, each raw message body is scanned individually (284-289), then after appending the XML-escaped full input string the same content is scanned again (298-303). The per-body scan added by this refactor already covers each body before escaping; the post-concatenation scan now re-runs O(total_input_bytes) work through both injection_scanner and leak_detector. At the 256 KiB cap this doubles scanning CPU. Pre-refactor code did only the full-string scan; new code does both. One is redundant.

**Fix:** Keep only the per-body pre-escape scan (drop the post-concatenation scan), or keep only the post-concatenation scan and remove the per-body scan — not both. If the escaped-form post-scan is intentional defense-in-depth, document the threat it catches that the per-body scan misses.

Also flagged by: maintainability/Low

File: crates/ironclaw_loop_support/src/compaction_task.rs [hunk 2]
```diff
@@ -400,16 +463,34 @@ fn message_kind_name(kind: MessageKind) -> &'static str {
...
+            '<' => "&lt;",
+            '>' => "&gt;",
+            _ => character.encode_utf8(&mut buffer),
+        };
+        push_checked(output, segment, cap)?;
+    }
+    Ok(())
+}
+
+fn push_checked(output: &mut String, segment: &str, cap: usize) -> Result<(), CompactionError> {
```
Comment (henrypark133) [inline crates/ironclaw_loop_support/src/compaction_task.rs:503]: **Low** — push_checked: no unit test for cap-at-exact-boundary vs one-over.

The refactor replaced a pre-computed escaped_message_len + single bounds check with per-segment push_checked calls. The existing contract test exercises rejection via a body well over the cap, but no test confirms boundary semantics: a serialized form exactly max_input_bytes should succeed, max_input_bytes+1 should return InputTooLarge. A future off-by-one to the > comparison would go undetected.

**Fix:** Add push_checked_accepts_payload_at_exact_cap_and_rejects_one_over covering cap == allowed and cap+1 == InputTooLarge.

File: crates/ironclaw_loop_support/src/compaction_task.rs [hunk 3]
```diff
@@ -368,24 +435,20 @@ fn compaction_message_body(message: &ThreadMessageRecord) -> Result<&str, Compac
...
+    kind: MessageKind,
+    body: &str,
+    cap: usize,
+) -> Result<(), CompactionError> {
+    push_checked(output, "<message sequence=\"", cap)?;
+    push_checked(output, &sequence.to_string(), cap)?;
+    push_checked(output, "\" kind=\"", cap)?;
+    push_checked(output, message_kind_name(kind), cap)?;
+    push_checked(output, "\">", cap)?;
+    append_escaped_xml_checked(output, body, cap)?;
```
Comment (henrypark133) [inline crates/ironclaw_loop_support/src/compaction_task.rs:454]: **Low** — append_escaped_xml_checked re-reads output.len() per character.

append_escaped_xml_checked iterates char-by-char and calls push_checked for every character, re-reading output.len() each iteration (O(N) reads/calls per body). The old escaped_xml_len + append pattern pre-computed total length in one pass with a single bounds check. Cold path (one compaction per session) so Low, but large Summary/System bodies near the 256 KiB cap add measurable overhead.

**Fix:** Batch non-special-character runs: scan ahead to the next special char, emit the run with one push_checked, then handle the special char. Or pre-compute escaped length and do one bounds check per message.

File: crates/ironclaw_loop_support/src/compaction_task.rs [hunk 4]
```diff
@@ -199,70 +225,94 @@ where
...
+            end_sequence: request.drop_through_seq,
+            messages: validated_messages,
+        })
+    }
+
+    fn build_input(
+        &self,
+        range: &ValidatedCompactionRange,
+    ) -> Result<CompactionInput, CompactionError> {
+        let mut text = String::new();
```
Comment (henrypark133) [inline crates/ironclaw_loop_support/src/compaction_task.rs:279]: **Low** — New method signatures use full path for already-imported module type.

run_inference and sanitize_summary reference ironclaw_turns::run_profile::SystemInferenceResponse by fully-qualified path, but the file already imports the rest of run_profile via a use block. Every other run_profile type in this file uses the import; SystemInferenceResponse is the outlier.

**Fix:** Add SystemInferenceResponse to the existing use ironclaw_turns::run_profile::{...} block and use the short name in both signatures.

Also flagged by: local-patterns/Low

File: crates/ironclaw_loop_support/src/compaction_task.rs [hunk 5]
```diff
@@ -69,6 +69,30 @@ pub(crate) struct CompactionTaskRequest {
...
+    messages: Vec<ValidatedCompactionMessage>,
+}
+
+struct ValidatedCompactionMessage {
+    sequence: u64,
+    kind: MessageKind,
+    body: String,
+}
+
+struct CompactionInput {
```
Comment (henrypark133) [inline crates/ironclaw_loop_support/src/compaction_task.rs:85]: **Low** — CompactionInput.bytes is always text.len() — redundant field.

CompactionInput carries text: String and bytes: usize where bytes is invariably text.len() (line 304). The struct is created once and consumed once; input_bytes is extracted immediately in run() before input.text is moved into inference. The field is always derivable from the string it accompanies.

**Fix:** Remove the bytes field; compute input_bytes as input.text.len() before passing input to run_inference.

Also flagged by: local-patterns/Nit

File: crates/ironclaw_loop_support/src/compaction_task.rs [hunk 6]
```diff
@@ -69,6 +69,30 @@ pub(crate) struct CompactionTaskRequest {
...
+    text: String,
+    bytes: usize,
+}
+
+struct SanitizedSummary {
+    content: String,
+    compression_ratio_ppm: u32,
```
Comment (henrypark133) [inline crates/ironclaw_loop_support/src/compaction_task.rs:91]: **Low** — compression_ratio_ppm: zero input_bytes path has no test.

compression_ratio_ppm returns 0 when input_bytes == 0 to avoid division by zero. This guard predates the refactor but remains in the changed file with no unit test; removal or branch change would go uncaught.

**Fix:** Add compression_ratio_ppm_returns_zero_for_empty_input covering input_bytes == 0.

**Review by henrypark133 — COMMENTED** [review]
## Code Review (multi-agent)

**Intent:** Refactor Reborn compaction task internals into typed private stages while keeping public port boundaries unchanged.

**Stats:** 6 findings (from 9 raw, 6 after dedup) across 1 file. Reviewers run: security, bugs, performance, tests, conventions, local-patterns, maintainability, pattern-refactor. Reviewers failed: none. Body-only: 0.

Clean behavior-preserving refactor — security, performance, and pattern-refactor found nothing. One Medium worth a look (redundant double-scan), rest are Low test/style nits. No correctness regressions.

### Bugs
1. **Medium** Double injection/leak scan — per-body scan then full-concatenated scan (`crates/ironclaw_loop_support/src/compaction_task.rs:277-306`, confidence 72) — anchor: compaction_task.rs:298. This refactor adds a per-body pre-escape scan; the pre-existing post-concatenation scan now re-runs O(total_input_bytes) over both scanners, doubling scan CPU at the 256 KiB cap. Keep one, or document the escaped-form post-scan's distinct threat model. Also flagged by maintainability/Low.
2. **Low** `append_escaped_xml_checked` re-reads `output.len()` per character (`compaction_task.rs:466-482`, confidence 65) — anchor: compaction_task.rs:466. Char-by-char `push_checked` is O(N) per body; batch non-special runs. Cold path, minor.

### Tests
3. **Low** `push_checked` has no cap-at-exact-boundary vs one-over test (`compaction_task.rs:484-494`, confidence 75) — anchor: compaction_task.rs:486. The new per-segment cap path is only exercised by a well-over body; add an exact-boundary test to lock the `>` comparison against off-by-one.
4. **Low** `compression_ratio_ppm` zero-input guard untested (`compaction_task.rs:518-526`, confidence 50) — anchor: compaction_task.rs:519.

### Conventions
5. **Low** New method signatures use full path for already-imported `SystemInferenceResponse` (`compaction_task.rs:308-340`, confidence 65) — anchor: CLAUDE.md code style. Add it to the existing `use ironclaw_turns::run_profile::{...}` block. Also flagged by local-patterns/Low.

### Maintainability
6. **Low** `CompactionInput.bytes` is always `text.len()` — redundant field (`compaction_task.rs:86-89`, confidence 65) — anchor: compaction_task.rs:304. Drop it; compute `input_bytes` at the use site. Also flagged by local-patterns/Nit (naming).

File: crates/ironclaw_loop_support/src/compaction_task.rs [hunk 7]
```diff
@@ -199,70 +225,94 @@ where
...
                 return Err(CompactionError::InvalidCutPoint);
             }
-            let body = compaction_message_body(message)?;
+            let body = compaction_message_body(&message)?.to_string();
+            validated_messages.push(ValidatedCompactionMessage {
+                sequence: message.sequence,
+                kind: message.kind,
+                body,
+            });
+        }
```
[the remark this answers is not in the document]
Comment (henrypark133) [reply crates/ironclaw_loop_support/src/compaction_task.rs:264]: Addressed in 8a7f31507: validate_range now takes ownership of message.content from the owned ThreadMessageRecord and still returns InvalidCutPoint when content is missing.

File: crates/ironclaw_loop_support/src/compaction_task.rs (hunk 1 above)
Comment (henrypark133) [reply crates/ironclaw_loop_support/src/compaction_task.rs:275]: Addressed in 8a7f31507 by documenting the second scan rather than removing it. The raw per-message scan remains the primary guard, and the serialized-input scan is intentionally retained to cover the exact model-visible escaped/delimited shape.

File: crates/ironclaw_loop_support/src/compaction_task.rs (hunk 2 above)
Comment (henrypark133) [reply crates/ironclaw_loop_support/src/compaction_task.rs:503]: Addressed in 8a7f31507: added push_checked_accepts_exact_cap_and_rejects_one_over for the exact-cap and one-over boundary.

File: crates/ironclaw_loop_support/src/compaction_task.rs (hunk 3 above)
Comment (henrypark133) [reply crates/ironclaw_loop_support/src/compaction_task.rs:454]: Addressed in 8a7f31507: append_escaped_xml_checked now batches contiguous non-special text runs and still checks the byte cap before each append.

File: crates/ironclaw_loop_support/src/compaction_task.rs (hunk 4 above)
Comment (henrypark133) [reply crates/ironclaw_loop_support/src/compaction_task.rs:279]: Addressed in 8a7f31507: SystemInferenceResponse is now imported in the run_profile use block and used directly in the method signatures.

File: crates/ironclaw_loop_support/src/compaction_task.rs (hunk 5 above)
Comment (henrypark133) [reply crates/ironclaw_loop_support/src/compaction_task.rs:85]: Addressed in 8a7f31507: removed CompactionInput.bytes and compute input.text.len() before moving the input into inference.

File: crates/ironclaw_loop_support/src/compaction_task.rs (hunk 6 above)
Comment (henrypark133) [reply crates/ironclaw_loop_support/src/compaction_task.rs:91]: Addressed in 8a7f31507: added compression_ratio_ppm_returns_zero_for_empty_input.
